File: day23/main.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var grid [][]byte
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, []byte(scanner.Text()))
	}

	height := len(grid)
	width := len(grid[0])
	start := Pos{
		x: 0,
		y: 0,
	}
	start.x = bytes.IndexByte(grid[0], '.')
	if start.x < 0 {
		log.Fatalln("No start")
	}
	end := Pos{
		x: 0,
		y: height - 1,
	}
	end.x = bytes.IndexByte(grid[height-1], '.')
	if end.x < 0 {
		log.Fatalln("No end")
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	undirectedGraph, directedGraph := contractPaths(start, end, grid, width, height)
	startID := start.y*width + start.x
	endID := end.y*width + end.x
	startID, startCost := findBranch(startID, undirectedGraph)
	endID, endCost := findBranch(endID, undirectedGraph)
	prunedCost := startCost + endCost
	fmt.Println("Part 1:", prunedCost+searchLongestGraph(startID, endID, directedGraph, width, height))
	fmt.Println("Part 2:", prunedCost+searchLongestGraph(startID, endID, undirectedGraph, width, height))
}

func searchLongestGraph(start, end int, graph map[int]map[int]int, w, h int) int {
	maxG := -1
	closedSet := make([]bool, w*h)
	openSet := NewStack[State]()
	openSet.Write(State{
		id:    start,
		g:     0,
		depth: 0,
	})
	curPath := NewStack[State]()
	for {
		cur, ok := openSet.Read()
		if !ok {
			break
		}

		if cur.id == end {
			maxG = max(maxG, cur.g)
			continue
		}

		for curPath.Len() > cur.depth {
			k, ok := curPath.Read()
			if !ok {
				log.Fatalln("cur path in bad state")
			}
			closedSet[k.id] = false
		}
		curPath.Write(cur)
		closedSet[cur.id] = true

		for k, v := range graph[cur.id] {
			if closedSet[k] {
				continue
			}
			openSet.Write(State{
				id:    k,
				g:     cur.g + v,
				depth: cur.depth + 1,
			})
		}
	}
	return maxG
}

func findBranch(pos int, graph map[int]map[int]int) (int, int) {
	edges := graph[pos]
	if len(edges) != 1 {
		return pos, 0
	}
	for k, v := range edges {
		return k, v
	}
	log.Fatalln("Unreachable")
	return -1, -1
}

func contractPaths(start, end Pos, grid [][]byte, w, h int) (undirected, directed map[int]map[int]int) {
	contractedGraph := map[int]map[int]int{}
	contractedDirectedGraph := map[int]map[int]int{}
	closedSet := make([]bool, w*h)
	openSet := NewRing[Pos](w * h)
	openSet.Write(start)
	exploredSet := make([]bool, w*h)
	exploredSet[start.y*w+start.x] = true
	var edges [4]Edge
	for {
		cur, ok := openSet.Read()
		if !ok {
			break
		}

		curKey := cur.y*w + cur.x
		closedSet[curKey] = true
		n := getPathEdges(cur, end, grid, w, h, closedSet, edges[:])
		closedSet[curKey] = false

		if cur != start && n < 2 && edges[0].pos != end {
			// eliminate dead ends
			continue
		}

		{
			var graphEdges map[int]int
			if m, ok := contractedGraph[curKey]; ok {
				graphEdges = m
			} else {
				graphEdges = map[int]int{}
				contractedGraph[curKey] = graphEdges
			}
			for _, o := range edges[:n] {
				key := o.pos.y*w + o.pos.x
				if o.pos != end && !exploredSet[key] {
					openSet.Write(o.pos)
					exploredSet[key] = true
				}
				if v, ok := graphEdges[key]; ok {
					if o.cost > v {
						graphEdges[key] = o.cost
						contractedGraph[key][curKey] = o.cost
					}
				} else {
					graphEdges[key] = o.cost
					var revGraphEdges map[int]int
					if m, ok := contractedGraph[key]; ok {
						revGraphEdges = m
					} else {
						revGraphEdges = map[int]int{}
						contractedGraph[key] = revGraphEdges
					}
					revGraphEdges[curKey] = o.cost
				}
			}
		}
		{
			var graphEdges map[int]int
			if m, ok := contractedDirectedGraph[curKey]; ok {
				graphEdges = m
			} else {
				graphEdges = map[int]int{}
				contractedDirectedGraph[curKey] = graphEdges
			}
			for _, o := range edges[:n] {
				key := o.pos.y*w + o.pos.x
				if o.forward {
					if v, ok := graphEdges[key]; ok {
						if o.cost > v {
							graphEdges[key] = o.cost
						}
					} else {
						graphEdges[key] = o.cost
					}
				}
				if o.rev {
					var revGraphEdges map[int]int
					if m, ok := contractedDirectedGraph[key]; ok {
						revGraphEdges = m
					} else {
						revGraphEdges = map[int]int{}
						contractedDirectedGraph[key] = revGraphEdges
					}
					if v, ok := revGraphEdges[curKey]; ok {
						if o.cost > v {
							revGraphEdges[curKey] = o.cost
						}
					} else {
						revGraphEdges[curKey] = o.cost
					}
				}
			}
		}
	}
	return contractedGraph, contractedDirectedGraph
}

type (
	State struct {
		id    int
		g     int
		depth int
	}

	Edge struct {
		pos     Pos
		cost    int
		forward bool
		rev     bool
	}

	Pos struct {
		x, y int
	}
)

func getPathEdges(pos, end Pos, grid [][]byte, w, h int, closedSet []bool, res []Edge) int {
	var unwindQueue []int
	count := getEdges(pos, grid, w, h, closedSet, res)
	var edges [4]Edge
	for idx, o := range res[:count] {
		cur := o
		for {
			if cur.pos == end {
				for _, i := range unwindQueue {
					closedSet[i] = false
				}
				unwindQueue = unwindQueue[:0]
				res[idx] = cur
				break
			}
			key := cur.pos.y*w + cur.pos.x
			closedSet[key] = true
			unwindQueue = append(unwindQueue, key)
			n := getEdges(cur.pos, grid, w, h, closedSet, edges[:])
			if n != 1 {
				for _, i := range unwindQueue {
					closedSet[i] = false
				}
				unwindQueue = unwindQueue[:0]
				res[idx] = cur
				break
			}
			first := edges[0]
			cur = Edge{
				pos:     first.pos,
				cost:    cur.cost + first.cost,
				forward: cur.forward && first.forward,
				rev:     cur.rev && first.rev,
			}
		}
	}
	return count
}

func getEdges(pos Pos, grid [][]byte, w, h int, closedSet []bool, res []Edge) int {
	count := 0
	cur := grid[pos.y][pos.x]
	{
		v := pos
		v.y--
		if isInBounds(v, w, h) && !closedSet[v.y*w+v.x] {
			if b := grid[v.y][v.x]; b != '#' {
				res[count] = Edge{
					pos:     v,
					cost:    1,
					forward: cur == '.' || cur == '^',
					rev:     b == '.' || b == 'v',
				}
				count++
			}
		}
	}
	{
		v := pos
		v.x--
		if isInBounds(v, w, h) && !closedSet[v.y*w+v.x] {
			if b := grid[v.y][v.x]; b != '#' {
				res[count] = Edge{
					pos:     v,
					cost:    1,
					forward: cur == '.' || cur == '<',
					rev:     b == '.' || b == '>',
				}
				count++
			}
		}
	}
	{
		v := pos
		v.y++
		if isInBounds(v, w, h) && !closedSet[v.y*w+v.x] {
			if b := grid[v.y][v.x]; b != '#' {
				res[count] = Edge{
					pos:     v,
					cost:    1,
					forward: cur == '.' || cur == 'v',
					rev:     b == '.' || b == '^',
				}
				count++
			}
		}
	}
	{
		v := pos
		v.x++
		if isInBounds(v, w, h) && !closedSet[v.y*w+v.x] {
			if b := grid[v.y][v.x]; b != '#' {
				res[count] = Edge{
					pos:     v,
					cost:    1,
					forward: cur == '.' || cur == '>',
					rev:     b == '.' || b == '<',
				}
				count++
			}
		}
	}
	return count
}

func isInBounds(pos Pos, w, h int) bool {
	return pos.x >= 0 && pos.y >= 0 && pos.x < w && pos.y < h
}

type (
	Stack[T any] struct {
		buf []T
	}
)

func NewStack[T any]() *Stack[T] {
	return &Stack[T]{}
}

func (b *Stack[T]) Write(m T) {
	b.buf = append(b.buf, m)
}

func (b *Stack[T]) Read() (T, bool) {
	if len(b.buf) == 0 {
		var v T
		return v, false
	}
	top := len(b.buf) - 1
	m := b.buf[top]
	b.buf = b.buf[:top]
	return m, true
}

func (b *Stack[T]) Peek() (T, bool) {
	if len(b.buf) == 0 {
		var v T
		return v, false
	}
	top := len(b.buf) - 1
	m := b.buf[top]
	return m, true
}

func (b *Stack[T]) Len() int {
	return len(b.buf)
}

type (
	Ring[T any] struct {
		buf []T
		r   int
		w   int
	}
)

func NewRing[T any](size int) *Ring[T] {
	if size < 2 {
		size = 2
	}
	return &Ring[T]{
		buf: make([]T, size),
		r:   0,
		w:   0,
	}
}

func (b *Ring[T]) resize() {
	next := make([]T, len(b.buf)*2)
	if b.r == b.w {
		b.w = 0
	} else if b.r < b.w {
		b.w = copy(next, b.buf[b.r:b.w])
	} else {
		p := copy(next, b.buf[b.r:])
		q := 0
		if b.w > 0 {
			q = copy(next[p:], b.buf[:b.w])
		}
		b.w = p + q
	}
	b.buf = next
	b.r = 0
}

func (b *Ring[T]) Write(m T) {
	next := (b.w + 1) % len(b.buf)
	if next == b.r {
		b.resize()
		b.Write(m)
		return
	}
	b.buf[b.w] = m
	b.w = next
}

func (b *Ring[T]) Read() (T, bool) {
	if b.r == b.w {
		var v T
		return v, false
	}
	next := (b.r + 1) % len(b.buf)
	m := b.buf[b.r]
	b.r = next
	return m, true
}

func (b *Ring[T]) Peek() (T, bool) {
	if b.r == b.w {
		var v T
		return v, false
	}
	m := b.buf[b.r]
	return m, true
}

File: day23/src/main.rs
use std::collections::hash_map::Entry;
use std::collections::{HashMap, VecDeque};
use std::fs::File;
use std::hash::Hash;
use std::io::{prelude::*, BufReader};

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let grid = reader
        .lines()
        .map(|line| Ok::<_, Box<dyn std::error::Error>>(line?.into_bytes()))
        .collect::<Result<Vec<_>, _>>()?;

    let h = grid.len();
    let w = grid[0].len();

    let start = Pos {
        x: grid[0].iter().position(|&v| v == b'.').ok_or("No start")?,
        y: 0,
    };
    let end = Pos {
        x: grid[h - 1]
            .iter()
            .position(|&v| v == b'.')
            .ok_or("No end")?,
        y: h - 1,
    };

    let (undirected_graph, directed_graph) = contract_paths(start, end, &GridGraph { w, h, grid });

    println!(
        "Part 1: {}\nPart 2: {}",
        directed_graph.cost + search_longest_path(&directed_graph),
        undirected_graph.cost + search_longest_path(&undirected_graph)
    );

    Ok(())
}

struct ContractedGraph {
    start: usize,
    end: usize,
    graph: Vec<Vec<(usize, usize)>>,
    cost: usize,
}

#[derive(Clone, Copy)]
struct Node<T> {
    value: T,
    g: usize,
    depth: usize,
}

fn search_longest_path(graph: &ContractedGraph) -> usize {
    let mut max_g = 0;
    let mut closed_set = VecBitSet::new(graph.graph.len());
    let mut open_set = Vec::new();
    let mut cur_path = Vec::new();
    open_set.push(Node {
        value: graph.start,
        g: 0,
        depth: 0,
    });
    while let Some(cur) = open_set.pop() {
        if cur.value == graph.end {
            max_g = max_g.max(cur.g);
            continue;
        }

        while cur_path.len() > cur.depth {
            let v: Node<_> = cur_path.pop().unwrap();
            closed_set.remove(v.value);
        }
        closed_set.insert(cur.value);
        cur_path.push(cur);

        for &(i, g) in &graph.graph[cur.value] {
            if closed_set.contains(i) {
                continue;
            }
            open_set.push(Node {
                value: i,
                g: cur.g + g,
                depth: cur.depth + 1,
            });
        }
    }
    max_g
}

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
struct Pos {
    x: usize,
    y: usize,
}

fn contract_paths(start: Pos, end: Pos, graph: &GridGraph) -> (ContractedGraph, ContractedGraph) {
    let mut node_names = NameIDMap::new();
    let start_id = node_names.get(start.y * graph.w + start.x);
    let end_id = node_names.get(end.y * graph.w + end.x);
    let mut contracted_graph = HashMap::new();
    let mut contracted_directed_graph = HashMap::new();

    let mut closed_set = VecBitSet::new(graph.w * graph.h);
    let mut open_set = VecDeque::new();
    let mut explored_set = VecBitSet::new(graph.w * graph.h);
    open_set.push_back(start);
    explored_set.insert(start.y * graph.w + start.x);
    let mut edges = Vec::new();
    let mut candidate_edges = Vec::new();
    while let Some(cur) = open_set.pop_front() {
        let cur_key = cur.y * graph.w + cur.x;
        let cur_id = node_names.get(cur_key);
        closed_set.insert(cur_key);
        let num_edges = get_path_edges(
            cur,
            end,
            &graph,
            &mut edges,
            &mut candidate_edges,
            &mut closed_set,
        );
        if cur != start && num_edges < 2 && (edges.len() == 0 || edges[0].pos != end) {
            // eliminate dead ends
            continue;
        }
        while let Some(o) = edges.pop() {
            let key = o.pos.y * graph.w + o.pos.x;
            let id = node_names.get(key);
            if id != end_id && !explored_set.contains(key) {
                open_set.push_back(o.pos);
                explored_set.insert(key);
            }
            match contracted_graph.entry((cur_id, id)) {
                Entry::Occupied(mut v) => {
                    let v = v.get_mut();
                    if o.cost > *v {
                        *v = o.cost;
                    }
                }
                Entry::Vacant(v) => {
                    v.insert(o.cost);
                }
            }
            match contracted_graph.entry((id, cur_id)) {
                Entry::Occupied(mut v) => {
                    let v = v.get_mut();
                    if o.cost > *v {
                        *v = o.cost;
                    }
                }
                Entry::Vacant(v) => {
                    v.insert(o.cost);
                }
            }
            if o.forward {
                match contracted_directed_graph.entry((cur_id, id)) {
                    Entry::Occupied(mut v) => {
                        let v = v.get_mut();
                        if o.cost > *v {
                            *v = o.cost;
                        }
                    }
                    Entry::Vacant(v) => {
                        v.insert(o.cost);
                    }
                }
            }
            if o.rev {
                match contracted_directed_graph.entry((id, cur_id)) {
                    Entry::Occupied(mut v) => {
                        let v = v.get_mut();
                        if o.cost > *v {
                            *v = o.cost;
                        }
                    }
                    Entry::Vacant(v) => {
                        v.insert(o.cost);
                    }
                }
            }
        }
    }

    let num_nodes = node_names.len();
    let undirected_graph = contracted_graph.into_iter().fold(
        vec![Vec::new(); num_nodes],
        |mut acc, ((from, to), v)| {
            acc[from].push((to, v));
            acc
        },
    );
    let directed_graph = contracted_directed_graph.into_iter().fold(
        vec![Vec::new(); num_nodes],
        |mut acc, ((from, to), v)| {
            acc[from].push((to, v));
            acc
        },
    );
    let (contracted_start_id, start_cost) = find_branch(start_id, &undirected_graph);
    let (contracted_end_id, end_cost) = find_branch(end_id, &undirected_graph);
    (
        ContractedGraph {
            start: contracted_start_id,
            end: contracted_end_id,
            graph: undirected_graph,
            cost: start_cost + end_cost,
        },
        ContractedGraph {
            start: contracted_start_id,
            end: contracted_end_id,
            graph: directed_graph,
            cost: start_cost + end_cost,
        },
    )
}

fn find_branch(a: usize, graph: &Vec<Vec<(usize, usize)>>) -> (usize, usize) {
    let edges = &graph[a];
    if edges.len() != 1 {
        return (a, 0);
    }
    *edges.first().unwrap()
}

fn get_path_edges(
    start: Pos,
    end: Pos,
    graph: &GridGraph,
    edges: &mut Vec<GridPath>,
    candidate_edges: &mut Vec<GridPath>,
    closed_set: &mut VecBitSet,
) -> usize {
    let num_edges = graph.get_edges(start, closed_set, edges);
    for i in 0..edges.len() {
        let mut cur = edges[i];
        loop {
            if cur.pos == end {
                edges[i] = cur;
                break;
            }
            if graph.get_edges(cur.pos, closed_set, candidate_edges) != 2 {
                candidate_edges.clear();
                edges[i] = cur;
                break;
            }
            let key = cur.pos.y * graph.w + cur.pos.x;
            closed_set.insert(key);
            // at least one edge must be present since nodes with two edges are always traversed
            let first = candidate_edges.pop().unwrap();
            cur = GridPath {
                pos: first.pos,
                cost: cur.cost + first.cost,
                forward: cur.forward && first.forward,
                rev: cur.rev && first.rev,
            };
        }
    }
    num_edges
}

struct GridGraph {
    w: usize,
    h: usize,
    grid: Vec<Vec<u8>>,
}

#[derive(Clone, Copy)]
struct GridPath {
    pos: Pos,
    cost: usize,
    forward: bool,
    rev: bool,
}

impl GridGraph {
    fn get_edges(&self, a: Pos, closed_set: &VecBitSet, res: &mut Vec<GridPath>) -> usize {
        let cur = self.grid[a.y][a.x];
        let mut num_edges = 0;
        let mut pos = a;
        pos.y = pos.y.wrapping_sub(1);
        if in_bounds(pos, self.w, self.h) {
            let b = self.grid[pos.y][pos.x];
            if b != b'#' {
                num_edges += 1;
                let key = pos.y * self.w + pos.x;
                if !closed_set.contains(key) {
                    res.push(GridPath {
                        pos,
                        cost: 1,
                        forward: cur == b'.' || cur == b'^',
                        rev: b == b'.' || b == b'v',
                    });
                }
            }
        }
        let mut pos = a;
        pos.x += 1;
        if in_bounds(pos, self.w, self.h) {
            let b = self.grid[pos.y][pos.x];
            if b != b'#' {
                num_edges += 1;
                let key = pos.y * self.w + pos.x;
                if !closed_set.contains(key) {
                    res.push(GridPath {
                        pos,
                        cost: 1,
                        forward: cur == b'.' || cur == b'>',
                        rev: b == b'.' || b == b'<',
                    });
                }
            }
        }
        let mut pos = a;
        pos.y += 1;
        if in_bounds(pos, self.w, self.h) {
            let b = self.grid[pos.y][pos.x];
            if b != b'#' {
                num_edges += 1;
                let key = pos.y * self.w + pos.x;
                if !closed_set.contains(key) {
                    res.push(GridPath {
                        pos,
                        cost: 1,
                        forward: cur == b'.' || cur == b'v',
                        rev: b == b'.' || b == b'^',
                    });
                }
            }
        }
        let mut pos = a;
        pos.x = pos.x.wrapping_sub(1);
        if in_bounds(pos, self.w, self.h) {
            let b = self.grid[pos.y][pos.x];
            if b != b'#' {
                num_edges += 1;
                let key = pos.y * self.w + pos.x;
                if !closed_set.contains(key) {
                    res.push(GridPath {
                        pos,
                        cost: 1,
                        forward: cur == b'.' || cur == b'<',
                        rev: b == b'.' || b == b'>',
                    });
                }
            }
        }
        num_edges
    }
}

fn in_bounds(a: Pos, w: usize, h: usize) -> bool {
    a.x < w && a.y < h
}

struct VecBitSet {
    bits: Vec<usize>,
    size: usize,
}

#[allow(dead_code)]
impl VecBitSet {
    fn new(size: usize) -> Self {
        Self {
            bits: vec![0; (size + (usize::BITS as usize - 1)) / usize::BITS as usize],
            size: 0,
        }
    }

    fn len(&self) -> usize {
        self.size
    }

    fn zero(&mut self) {
        for i in self.bits.iter_mut() {
            *i = 0;
        }
        self.size = 0;
    }

    fn contains(&self, i: usize) -> bool {
        let a = i / usize::BITS as usize;
        let b = i % usize::BITS as usize;
        let mask = 1 << b;
        self.bits[a] & mask != 0
    }

    fn insert(&mut self, i: usize) {
        let a = i / usize::BITS as usize;
        let b = i % usize::BITS as usize;
        let mask = 1 << b;
        if self.bits[a] & mask == 0 {
            self.bits[a] |= mask;
            self.size += 1;
        }
    }

    fn remove(&mut self, i: usize) {
        let a = i / usize::BITS as usize;
        let b = i % usize::BITS as usize;
        let mask = 1 << b;
        if self.bits[a] & mask != 0 {
            self.bits[a] &= !mask;
            self.size -= 1;
        }
    }
}

struct NameIDMap<T> {
    names: HashMap<T, usize>,
}

impl<T: Eq + Hash> NameIDMap<T> {
    fn new() -> Self {
        Self {
            names: HashMap::new(),
        }
    }

    fn len(&self) -> usize {
        self.names.len()
    }

    fn get(&mut self, a: T) -> usize {
        let next = self.names.len();
        *self.names.entry(a).or_insert(next)
    }
}
